### PR TITLE
initial version, need to add comments, more tests and check edge cases

### PR DESCRIPTION
**Pull request by nchiki**

### CODE REVIEW

File: nchiki/assignment-3/Grid.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+
+public class Grid {
+
+  public char[][] grid;
+  public HashMap<Character, List<Character>> adjacencyMap;
```
Comment (TDiva) [inline nchiki/assignment-3/Grid.java]: This assumes that all characters in grid are unique, which might not be true.

File: nchiki/assignment-3/LexiconTests.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import org.junit.Test;
+
+import java.util.List;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.*;
+
+public class LexiconTests {
+
+  List<String> words =
```
Comment (TDiva) [inline nchiki/assignment-3/LexiconTests.java]: It's nicer to have data you test on excplicitly created in test case - if tests are long you can easier find what the data is. Or at least in special SetUp method.

File: nchiki/assignment-3/LexiconTests.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  Lexicon lexicon = new Lexicon(words);
+
+  List<String> sampleWords = List.of("CAR", "CARD", "CART", "CAT");
+  Lexicon sample = new Lexicon(sampleWords);
+
+  char[][] mat = {{'S', 'A', 'R'}, {'T', 'C', 'D'}};
+  Grid grid = new Grid(mat);
+
+  @Test
+  public void returnsWhetherWordIsInLexicon() {
```
Comment (TDiva) [inline nchiki/assignment-3/LexiconTests.java]: This is better to split into 2 test cases: testStringIsInLexicon and testStringIsNotInLexicon.
Also please add more test cases - empty stirngs, non-latin letters, characters, etc.
A grid with duplicate characters as well please!

File: nchiki/assignment-3/Grid.java (hunk 1 above)
Comment (TDiva) [reply nchiki/assignment-3/Grid.java]: Please, replace this with correct representation of a graph, grid may contain duplicative characters.

File: nchiki/assignment-3/GridTrie.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+public class GridTrie {
+
+  int ALPHABET = 26;
+  private final char[][] grid;
+  List<String> entries = new LinkedList<>();
+
+  static class TrieNode {
+    private int ALPHABET = 26;
+    TrieNode[] children = new TrieNode[ALPHABET];
```
Comment (TDiva) [inline nchiki/assignment-3/GridTrie.java]: Again: there might be not only english characters used.
Please, consider using HashMap for such cases.

File: nchiki/assignment-3/GridTrie.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    }
+  }
+
+  private void traverseChildren(
+      TrieNode root, int row, int col, boolean visited[][], String curString, List<String> words) {
+    for (int i = 0; i < ALPHABET; i++) {
+      if (root.children[i] != null) {
+        char curChar = (char) (i + 'A');
+
+        if (inBounds(row + 1, col + 1, visited) && grid[row + 1][col + 1] == curChar) {
```
Comment (TDiva) [inline nchiki/assignment-3/GridTrie.java]: Looks like this can be replaced by loop, exactly the same operation is done 8 times.

File: nchiki/assignment-3/GridTrie.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    boolean[][] visited = new boolean[rows][cols];
+    String curString = "";
+    List<String> words = new LinkedList<>();
+
+    TrieNode curNode = root;
+    for (int i = 0; i < rows; i++) {
+      for (int j = 0; j < cols; j++) {
+        if (curNode.children[(this.grid[i][j])-'A'] != null) {
+          curString += grid[i][j];
+          findWord(curNode.children[(grid[i][j]) - 'A'], visited, i, j, curString, words);
```
Comment (TDiva) [inline nchiki/assignment-3/GridTrie.java]: this is also case sensitive and might fail with lower case letters.

File: nchiki/assignment-3/GridTrie.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    return words;
+  }
+
+  private boolean inBounds(int row, int col, boolean[][] visited) {
+    return (row >= 0
+        && row < grid.length
+        && col >= 0
+        && col < grid[0].length
+        && !visited[row][col]);
```
Comment (TDiva) [inline nchiki/assignment-3/GridTrie.java]: I think name of function is misleading - inBounds doesn't say anything about checking is the cell visited or not.

File: nchiki/assignment-3/GridTrieTests.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    private void initTrie() {
+        lex.insert("CAR");
+        lex.insert("CARD");
+        lex.insert("CART");
+        lex.insert("CAT");
+    }
+
+    @Test
+    public void returnsLexiconWordsInGrid() {
```
Comment (TDiva) [inline nchiki/assignment-3/GridTrieTests.java]: More tests please!

File: nchiki/assignment-3/Grid.java [hunk 9]
```diff
@@ -1,70 +1,4 @@
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-
 public class Grid {
```
Comment (TDiva) [inline nchiki/assignment-3/Grid.java]: This class is empty, why would we need it?

File: nchiki/assignment-3/LexiconTrie.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class LexiconTrie {
+
+  int ALPHABET = 94;
```
Comment (TDiva) [inline nchiki/assignment-3/LexiconTrie.java]: Using magic constants is also highly discouraged. In this case you can replace array of trie nodes with map.
It's also highly unclear where number 94 comes from :)

File: nchiki/assignment-3/LexiconTrie.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+     */
+  private void traverseChildren(
+      TrieNode root, int row, int col, boolean visited[][], String curString, List<String> words) {
+    for (int i = 0; i < ALPHABET; i++) {
+      if (root.children[i] != null) {
+        char curChar = (char) (i + BEGINNING_ALPHABET);
+
+        for (int j = -1; j <= 1; j++) {
+          if ( j != 0) {
+            if (isUnvisited(row + j, col + j, visited) && grid[row + j][col + j] == curChar) {
```
Comment (TDiva) [inline nchiki/assignment-3/LexiconTrie.java]: There are still 5 times repeated code, and I believe it checks some cells twice.
Please, restructure it.

File: nchiki/assignment-3/LexiconTrie.java [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+      }
+    }
+    return words;
+  }
+
+
+    /**
+     * returns true if given coordinates are in bounds of the grid and the cell was not visited yet
+     */
+  private boolean isUnvisited(int row, int col, boolean[][] visited) {
```
Comment (TDiva) [inline nchiki/assignment-3/LexiconTrie.java]: Usually it's better to make functions simpler in naming and behavior, isVisited = !isUnvisited, but name is much clearer. Also, it's good to divide it into two - it's not very clear why isUnvisited checks also grid borders.
